[apiserver] Add migration doc from v1 to v2 (#3812)
ray-project/kuberay: apiserversdk/docs/migration.md

> # KubeRay APIServer v2 Migration Guide

Reviewer: ```suggestion
# KubeRay APIServer v2 Migration Plan
```

I think this is the last change we need on this doc.

## Overview

KubeRay APIServer v2 introduces a more maintainable, Kubernetes-native, and flexible interface for managing Ray clusters.

In v1, exposing new fields required modifying protobuf definitions, regenerating HTTP/gRPC clients, and updating
tests — a time-consuming and error-prone process that often delayed support for new features. This manual
synchronization between CRDs and protobuf definitions added significant maintenance overhead and slowed
developer velocity.

With v2, we eliminate these bottlenecks by directly reusing the OpenAPI schema defined in the Kubernetes CRDs.
Instead of manually defining fields or generating new clients, APIServer v2 acts as a transparent HTTP proxy to the
Kubernetes API server. All CRD fields are exposed by default, and advanced behaviors such as compute template
injection, default values, and mutations can be implemented using user-defined middleware functions (UDFs).

To simplify the system further, gRPC support has been removed in favor of HTTP-only APIs. This approach aligns
better with native Kubernetes tooling and improves extensibility, maintainability, and onboarding for

> infrastructure engineers. This guide outlines the major changes from APIServer v1 to v2 and provides guidance for migrating

Reviewer: ```suggestion
infrastructure engineers. This document outlines the major changes from APIServer v1 to v2.
```

# KubeRay APIServer v2 Migration Guide

## Target Customers

- Advanced and early adopters who currently use the v1 API in production
- Infrastructure and platform engineers who have customized or extended the KubeRay API Server for internal tooling or environments

## Overview

KubeRay API Server v2 introduces a more maintainable, Kubernetes-native, and flexible interface for managing Ray clusters.

In v1, exposing new fields required modifying protobuf definitions, regenerating HTTP/gRPC clients, and updating
tests — a time-consuming and error-prone process that often delayed support for new features. This manual
synchronization between CRDs and protobuf definitions added significant maintenance overhead and slowed
developer velocity.

With v2, we eliminate these bottlenecks by directly reusing the OpenAPI schema defined in the Kubernetes CRDs.
Instead of manually defining fields or generating new clients, v2 acts as a transparent HTTP proxy to the

> Kubernetes API server. All CRD fields are exposed by default, and advanced behaviors such as compute template

Author: Oh I see — I didn’t know that. Thanks for the explanation.

| Compute template support    | Requires hardcoded logic in API server                                       | Handled via HTTP middleware/UDF; official support starts in Stage 2                          |
| Pagination support          | Manual implementation required                                               | Natively supported using Kubernetes API pagination                                            |
| Documentation policy        | Partial or undocumented                                                      | Official documentation includes only v2                                                       |
| Maintenance and extensibility | High overhead: PR needed per new field, manual sync with operator         | Lightweight: CRD fields auto-reflected, extensible via middleware, 50% less maintenance       |

## Migration Plan

The migration from v1 to v2 will occur in three stages:

- Stage 1 – v2 released and validated
  - v2 becomes publicly available and feature-complete
  - All major features from v1 are supported

>   - gRPC support is not included in v2, but remains available in v1 during the depreciation period

Author: Thanks!
